python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,169 @@
+
+    Parameters
+    ----------
+    mixin_name : str
+        The name of the class. This argument should be the same as the object
+        that this function's output is assigned to, e.g.
+        :code:`Baz = _create_delegating_mixin("Baz", ...)`.
+    docstring : str
+        The documentation string for the mixin class.
+    category_name : str
+        The category of operations for which a mixin is being created. This
+        name will be used to define the following attributes as shown in the
+        example below:
+            - f'_{category_name}_DOCSTRINGS'
+            - f'_VALID_{category_name}S'
+            - f'_SUPPORTED_{category_name}S'
+    operation_name : str
+        The name given to the core function implementing this category of
+        operations.  The corresponding function is the entrypoint for child
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
+    ...     "MyFoo", "MyFoo's docstring", "FOO", "do_foo", {"foo1", "foo2"}
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         raise NotImplementedError
+
+    >>> # MyFoo can be used as follows.
+    >>> class BarImplementsFoo(MyFoo):
+    ...     _SUPPORTED_FOOS = ("foo1",)
+    ...     _FOO_DOCSTRINGS = {"ret": "42"}
+    ...
+    ...     # This method's docstring will be formatted and used for all valid
+    ...     # operations. The `op` key is always automatically filled in, while
+    ...     # other keys are formatted using _FOO_DOCSTRINGS.
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         '''Perform the operation {op}, which returns {ret}.'''
+    ...         return 42
+
+    >>> bar = BarImplementsFoo()
+    >>> print(bar.foo1())
+    42
+
+    >>> # This will raise an AttributeError because foo2 is not supported by
+    >>> # the BarImplementsFoo subclass of MyFoo.
+    >>> # print(bar.foo2())
+
+    >>> # The docstring is formatted with the operation name as well as any
+    >>> # additional keys provided via the _FOO_DOCSTRINGS parameter.
+    >>> print(bar.foo1.__doc__)
+    Perform the operation foo1, which returns 42.
+    """
+
+    docstring_attr = f"{category_name}_DOCSTRINGS"
+    validity_attr = f"_VALID_{category_name}S"
+
+    class OperationMixin:
+        @classmethod
+        def _add_operation(cls, operation):
+            # This function creates operations on-the-fly.
+
+            # Generate a signature without the `op` parameter.
+            base_operation = getattr(cls, operation_name)
```

Comment:
When debugger-ing through the code I was initially little confused by `operation_name`. Could we name it `operation_category_name` to distinguish it in this context from `operation`, which is a specific op?

Reply:
Good idea, done.